src/Accelerators/NNPA/Conversion/ONNXToZHigh/DevicePlacement.cpp

```diff
@@ -108,9 +120,18 @@ void DevicePlacementPass::runOnOperation() {
     StringAttr device = op->getAttrOfType<mlir::StringAttr>(DEVICE_ATTRIBUTE);
     if (device && !device.getValue().empty())
       return WalkResult::advance();
-    // Otherwise, set device.
-    if (!cpuOps.contains(op))
-      op->setAttr(DEVICE_ATTRIBUTE, StringAttr::get(context, NNPA_DEVICE));
+    // Op that is legal (should remain on the CPU) as determined by compiler
+    // analysis.
+    if (cpuOps.contains(op))
+      return WalkResult::advance();
+    // Now we have an operation that can work on the NNPA, check if its
+    // beneficial
+    if (useZHighCostModel && !isOpFasterOnNNPA(op, &dimAnalysis)) {
```

Comment:
@tungld Please let me know if that is how you envisioned this to be used. I placed it there because to the extend possible, I only want to invoke the `isOpFasterOnNNPA` for operations that I know are candidates... otherwise there might be a lot of ops that are irrelevant to being with. Open to suggestion if you had something else in mind.

Reply:
@AlexandreEichenberger I plan to do this earlier, at Line 74. The code here will respect what you add by the cost model (e.g. `device = cpu`). Something like this at Line74.
```C
 if (useZHighCostModel) {
   module.walk([&](Operation *op) {
    if (!isOpFasterOnNNPA(op, &dimAnalysis))
      op->setAttr(DEVICE_ATTRIBUTE, StringAttr::get(context, CPU_DEVICE));
    });
  }
```

Reply:
The legality check is aware of `device=cpu` in an operation if it is set, so the compiler will assign NNPA for the other ops only.

Reply:
It's possible, but it will see all of the onnx ops, as opposed to only the ops that were deemed to be target for NNPA. 
So it should run faster (seen fewer ops) and also report on only ops that qualify. That helps for printing debugging (only focusing on the important ops). So do you think its worth it to move the code before?

Reply:
Yes, it is faster here. Actually, the code here is to annotate ops based the result of analysis of rewriting patterns, so it's a bit confusing if we add the cost model here. So, we will annotate the ops using the cost model first, and the rewrite patterns are internally respect `device` attribute annotated by the cost model.

Reply:
@AlexandreEichenberger fyi, I am creating a PR to set device by using a json file: #2536. It works now but still needs to add some lit test.